### ISSUE TITLE
What are the features in your model?

### ISSUE THREAD

**Issue by Coding511** [xiaomi1024/code_SAMS#2]
Hi there, I am sorry, but in the paper, there is no detail about the features S and T? which acoustic features and which word embeddings have you used there? Precisely what is the shape of the feature vector? Can I reproduce the feature pickle files?

Regards

**Comment by Michell-H** [comment]
These features are common and easy to obtain. In the stage of feature extraction, we apply 26 Mel filters to extract log Mel-frequency spectral coefficients (including first derivative and second derivative) for audio features (S->(N,500,78), and N is the number of total samples ). We perform mean pooling with 5 steps and 4 steps to get the 500  frames for all audio samples. We preserve 150 and 100 words  length of all utterances on the IEMOCAP and MELD datasets using the pre-trained BERT model (uncased_L-12_H-768_A-12), respectively (T->(N, 100, 768) or (N, 150, 768)).

**Comment by dzr1026** [comment]
Hello, I downloaded the feature file you shared (IEMOCAP_ST.pkl) and found that the dimension of speech in the IEMOCAP dataset is (N, 500, 123), not (N,500,78) as you mentioned above, can you explain some further, thank you very much.

**Comment by Coding511** [comment]
@dzr1026 Yes. Can anyone plz refer a snippet to convert the feature vectors into such pickle file. Another question is why did the authors use joblib, I tried reading through pickle its unicode is readable.

**Comment by Coding511** [comment]
@dzr1026 @Michell-H  considering audio features, 

len(S[1][2][3])
Out[3]: 123

len(S[1][2])
Out[4]: 500

len(S[1])
Out[5]: 557

could you please explain the meaning of the above dimensions?
In the paper, it should be like this:    (number of frames * feature dimension)
Is 557 is the number of frames in a particular utterance? then what is 500, I am so confused. Please help.

**Comment by Michell-H** [comment]
> Hello, I downloaded the feature file you shared (IEMOCAP_ST.pkl) and found that the dimension of speech in the IEMOCAP dataset is (N, 500, 123), not (N,500,78) as you mentioned above, can you explain some further, thank you very much.

This data is a 123-dimensional speech feature containing 78-dimensional log Mel-frequency spectral coefficients features. In order to eliminate misunderstandings, we have re-uploaded the features again and eliminated irrelevant features that were not used, please refer to the link in the download.txt  file.

**Comment by Michell-H** [comment]
> @dzr1026 @Michell-H considering audio features,
> len(S[1][2][3]) Out[3]: 123
> len(S[1][2]) Out[4]: 500
> len(S[1]) Out[5]: 557
> could you please explain the meaning of the above dimensions? In the paper, it should be like this: (number of frames * feature dimension) Is 557 is the number of frames in a particular utterance? then what is 500, I am so confused. Please help.

123 (Actually it is 78, the new version has been updated)->Number of features  
500->Number of frames
557->This value is the number of samples of one speaker in Session1 in the IEMOCAP dataset.

**Comment by Coding511** [comment]
Thank you by the time I got it.
however, what is the reason for 500 frames? Are you saying from the whole utterance you are using 500 frames each of 25ms and 10 ms overlap? If yes, How does each frame only give a 26mel filter bank vector? (500,78)
As log mel spectrogram is 2D structure, right? So each frame of 25ms should give a 2D structure using librosa library. Please guide sir.

**Comment by Michell-H** [comment]
> Thank you by the time I got it. however, what is the reason for 500 frames? Are you saying from the whole utterance you are using 500 frames each of 25ms and 10 ms overlap? If yes, How does each frame only give a 26mel filter bank vector? (500,78) As log mel spectrogram is 2D structure, right? So each frame of 25ms should give a 2D structure using librosa library. Please guide sir.

500 frames is the number after pooling. For example, if the pooling step is 5, the sample duration that can be covered by a 25 ms window of 10 ms move operation for voice data is about 25s ((500*5*10+25)/1000). This duration should cover almost all samples.

**Comment by Coding511** [comment]
@Michell-H Unfortunately it is still not clear sir.
If you are stating that 25 sec is the average sample duration you are considering here, that's fine, but this pooling and above numbers ??? sorry please elaborate a little as if the features extraction step is not clear, we can't cite the work for comparison right?

Also in my last comment, I am trying to say for whatever reason you are considering 500 frames, log mel spectrogram is a 2-dimensional feature, so for every such 25ms frame (500 samples in it) it should have a 2d feature dimension.

If the above theory is wrong then an alternative is you are considering 16-sec length audio and for that mel spec features dimension is 500X78 (using the default setting of librosa).

Please help which one is correct here. A detailed answer would be appreciated as it is your work no one else can better explain it.

Thank you

**Comment by Coding511** [comment]
Also in Equation (4) what is the difference between suffix i and superscript m. Both are indexing I think, then why two vaiables sir?

**Comment by Coding511** [comment]
@Michell-H  are you available ? We can talk through email also.

**Comment by Michell-H** [comment]
First, we  apply 26 Mel filters to extract log Mel-frequency spectral coefficients (including first derivative and second derivative) with Hamming windows of 25 ms and a 10 ms shift (If a sample is x ms, the features of ((x-25)/10 + 1) frames can be obtained at this stage). Then, we perform mean pooling with 5 steps to get the final frames  for a utterance (this stage can get the features of ((x-25)/10 + 1)/5) frames). Finally, we fill all samples to 500 frames . As mentioned above, 500 frames is the result of 25s of signal feature extraction and is the longest sample frame length (((25000ms-25ms)/10+1)/5 approximately equal to 500).

**Comment by Michell-H** [comment]
> Also in Equation (4) what is the difference between suffix i and superscript m. Both are indexing I think, then why two vaiables sir?

m denotes the ordinal number of the modality, and i denotes the ordinal number of the frame or word.

**Comment by Coding511** [comment]
@Michell-H Thank you so much
Yes , I was taking i and m in same way, But you have written i=1 to T and T belongs to (la, lt) where la =500 and lt=150 right?
hence i=650 which is same as m.  I think T made things confusing.

**Comment by Michell-H** [comment]
> @Michell-H Thank you so much Yes , I was taking i and m in same way, But you have written i=1 to T and T belongs to (la, lt) where la =500 and lt=150 right? hence i=650 which is same as m. I think T made things confusing.

m and i mean different things, and they are not the same. i<=500(la) or i<=150(lt), but i  != 650. m=a or m=t is just a modal identifier.

**Comment by Coding511** [comment]
> @Michell-H Unfortunately it is still not clear sir. If you are stating that 25 sec is the average sample duration you are considering here, that's fine, but this pooling and above numbers ??? sorry please elaborate a little as if the features extraction step is not clear, we can't cite the work for comparison right?
> 
> Also in my last comment, I am trying to say for whatever reason you are considering 500 frames, log mel spectrogram is a 2-dimensional feature, so for every such 25ms frame (500 samples in it) it should have a 2d feature dimension.
> 
> If the above theory is wrong then an alternative is you are considering 16-sec length audio and for that mel spec features dimension is 500X78 (using the default setting of librosa).
> 
> Please help which one is correct here. A detailed answer would be appreciated as it is your work no one else can better explain it.
> 
> Thank you

@Michell-H dear author, I am still not getting how are you evaluating these feature vector. Above all this term pooling is not clear here.
You must be using librosa for the same, for a 25 sec signal and frame rate of 16000 the code is given below:

import librosa 
import numpy as np
path='E:/Ses01F_impro01_F000.wav'
win_length=400000 % for a 25sec signal=25x16000=400000
audio_data,sr  = librosa.load(path,sr=16000)
diff = win_length-len(audio_data)
create_arr = np.zeros([1,diff])
final_data  = np.concatenate((audio_data,create_arr[0]))
audio_data = final_data
A=librosa.feature.melspectrogram(audio_data,sr=sr,win_length=400) % win_length=25ms=400

A.shape
Out[20]: (128, 782)

Here by default number of filters are 128 but that I can change by passing argument n_mels=13, and 782 is through holp_length=512 , i.e. 400000/512+1=782

What else I need to change here.

**Comment by Michell-H** [comment]
We give the following example:

import numpy as np
import wave
import python_speech_features as ps

file = wave.open('xxxxxx.wav.', 'r') # change your wav file
params = file.getparams()
nchannels, sampwidth, framerate, wav_length = params[:4]
str_data = file.readframes(wav_length)
wavedata = np.fromstring(str_data, dtype=np.short)
log_f = ps.logfbank(wavedata, framerate,  winlen=0.025, winstep=0.01, nfilt=26, nfft=512, lowfreq=0, highfreq=None, preemph=0.97)
delta1 = ps.delta(log_f, 2)
delta2 = ps.delta(delta1, 2)
feat_ = np.concatenate((log_f, delta1, delta2), axis=1)

We perform a mean pooling operation on this feat_.

**Comment by Coding511** [comment]
@Michell-H Thank you so much, I don't know why I was extracting Mel spectrogram , you are taking Log mel filter banks.  
Now I can close this with this last query, This `feat_.shape` is [1599,78]. This 1599 is observed after making `wavdata=25sec` long.
So how this mean pooling is performed to make it 500,78. As  we mean pool across this dimension it should be 1599/5=320 not 500.
Thanks in advance.

**Comment by Michell-H** [comment]
> @Michell-H Thank you so much, I don't know why I was extracting Mel spectrogram , you are taking Log mel filter banks. Now I can close this with this last query, This `feat_.shape` is [1599,78]. This 1599 is observed after making `wavdata=25sec` long. So how this mean pooling is performed to make it 500,78. As we mean pool across this dimension it should be 1599/5=320 not 500. Thanks in advance.

Yes, so to keep the data dimensions the same, just perform 0 padding for less than 500 frames. We have performed some previous works and found that  log mel filter banks features exhibit better recognition performance than MFCC for speech emotion recognition.

**Comment by Coding511** [comment]
I think I cannot convey, but this feature extraction part is really fishy and tough to reproduce the result despite this simple model. In the above statement of mine, I have already padded `wave data` to make it 25 sec long, So if all files are already 25 sec long, then `log_f` feature would be of this dimension only [1599, 78], and it leads to 320 frames after 5 step pooling. Why this zero padding now, I mean why making all files padded with zeros.  Could you paste that one extra line where you are doing pooling also.

**Comment by Michell-H** [comment]
> I think I cannot convey, but this feature extraction part is really fishy and tough to reproduce the result despite this simple model. In the above statement of mine, I have already padded `wave data` to make it 25 sec long, So if all files are already 25 sec long, then `log_f` feature would be of this dimension only [1599, 78], and it leads to 320 frames after 5 step pooling. Why this zero padding now, I mean why making all files padded with zeros. Could you paste that one extra line where you are doing pooling also.

import numpy as np
import wave
import python_speech_features as ps

file = wave.open('1.wav.', 'r') # change your wav file
params = file.getparams()
nchannels, sampwidth, framerate, wav_length = params[:4]
str_data = file.readframes(wav_length)
wavedata = np.fromstring(str_data, dtype=np.short)
log_f = ps.logfbank(wavedata, framerate,  winlen=0.025, winstep=0.01, nfilt=26, nfft=512, lowfreq=0, highfreq=None, preemph=0.97)
delta1 = ps.delta(log_f, 2)
delta2 = ps.delta(delta1, 2)
feat_ = np.concatenate((log_f, delta1, delta2), axis=1)
# mean pooling
if len(feat_) % 5 != 0:
    for i in range(5 - (len(feat_) % 5)):
        feat_ = np.vstack([feat_, np.zeros(78)])
tmp = np.zeros([len(feat_) // 5, 78])
for i in range(len(feat_) // 5):
    tmp[i] = np.mean(np.split(feat_, len(feat_) / 5)[i], 0)
feat_pool = tmp
# padding
feat_pad = np.pad(feat_pool, ((0, 500 - feat_pool.shape[0]), (0, 0)), 'constant',  constant_values=0)

**Comment by AnuragIISC** [comment]
> Hello, I downloaded the feature file you shared (IEMOCAP_ST.pkl) and found that the dimension of speech in the IEMOCAP dataset is (N, 500, 123), not (N,500,78) as you mentioned above, can you explain some further, thank you very much.

Hello sir, could you please direct me where the features are uploaded?
